verification/spec/connection-handshake/chm.tla

```diff
@@ -0,0 +1,241 @@
+-------------------------------- MODULE CHM --------------------------------
+
+EXTENDS Naturals, FiniteSets
+
+
+CONSTANT MaxHeight     \* Maximum height of all the chains in the system.
+
+
+VARIABLES
+    chModule,   \* The state of this connection handshake module (chm).
+    turn,       \* Keep track of who takes a step: either chModule or environment.
+    inMsg,      \* A buffer holding any message incoming to the chModule.
+    outMsg      \* A buffer holding any message outbound from the chModule.
+
+
+vars == <<turn, chModule, inMsg, outMsg>>
+
+Heights == 1..MaxHeight
+
+ConnectionStates == {"UNINIT", "INIT", "TRYOPEN", "OPEN"}
+
+ConnectionIDs == {"connAtoB", "connBtoA"}
+ClientIDs == { "clientA", "clientB" }
+
+Null == "none"
+noMsg == [ type |-> "none" ]
+noErr == "errNone"
+
+
+(***************************** ConnectionEnds *****************************
+    A set of connection end records.
+    A connection end record contains the following fields:
+
+    - connectionID -- a connection identifier
+      Stores the connection identifier for this connection end.
+
+    - clientID -- a client identifier
+      Stores the client identifier associated with this connection end.
+
+    Optional, add later:
+    - prefix
+    
+ ***************************************************************************)
+ConnectionEnds ==
+    [
+        connectionID : ConnectionIDs \union {Null},
+        clientID : ClientIDs \union {Null}
+    ]
+
+
+(***************************** Connections *****************************
+    A set of connection records.
+    A connection record contains the following fields:
+
+    - localEnd -- a connection end
+      Stores the connection end for the local chain.
+
+    - remoteEnd -- a connection end
+      Stores the connection end for the remote chain.
+ ***************************************************************************)
+Connections ==
+    [
+        localEnd : ConnectionEnds,
+        remoteEnd : ConnectionEnds
+    ]
+
+
+(***************************** ConnectionHandshakeModules ******************
+    A set of records defining the CHM.
+    A CHM record contains the following fields:
+
+    - connectionState -- a string
+      Stores the current state of this connection. It has one of the
+      following values: "UNINIT", "INIT", "TRYOPEN", "OPEN".
+
+    - connection -- a connection with two ends
+      Stores the connection.
+
+    - chainHeight -- a height
+      Stores the height of the chain running alongside the CHM.
+ ***************************************************************************)
+ConnectionHandshakeModules ==
+    [
+        connectionState : ConnectionStates,
```

Comment:
why isn't `connectionState` part of `connection`?

Reply:
Initially, it was part of `connection`.

But then I noticed a `connection` will be part of the function handler signatures, as follows:

```
ConnectionHandshakeMessages ==
    [type : {"ConnOpenInit"}, 
     connection : Connections]
    \union
    [type : {"ConnOpenTry"},
        connection : Connections,
        stateProof : Proofs, 
        consensusHeight : Proofs]]
```

Based on this, I decided it would be more elegant to pull out the `connectionState` field from `Connection`, because this field is irrelevant (and represents purely local information). But this is a cosmetic choice. If this is confusing, I'll put it back. Especially if we want to model multiple connection handshakes in this spec, then it will be important to put the field back into `Connection`.

Reply:
maybe we should define a connection parameter.

Reply:
Hmm, I think I like that.

We would end up with the following abstractions:

```tla
ConnectionEnds ==
    [
        connectionID : ConnectionIDs \union {NullConnectionId},
        clientID : ClientIDs \union {NullClientId}
        \* commitmentPrefix add later
    ]

ConnectionParameters ==
    [
        localEnd : ConnectionEnds,
        remoteEnd : ConnectionEnds
    ]

Connections ==
    [
        parameters : ConnectionParameters,
        state : ConnectionState,
        \* any other fields that are specific for a connection
    ]
```

The field `ConnectionParameters` uniquely identifies a connection. And `ConnectionParameters` go as parameters into all `ConnectionHandshakeMessages`.